### PR TITLE
internal/ui: remove Cgo on darwin in ui_glfw_darwin.go

### PR DESCRIPTION
**Pull request by TotallyGamerJet**

### CODE REVIEW

**Comment by TotallyGamerJet** [comment]
@hajimehoshi please take a look

File: internal/ui/ui_glfw_darwin.go [hunk 1]
```diff
@@ -345,27 +270,76 @@ func (u *userInterfaceImpl) isNativeFullscreenAvailable() bool {
...
 }
 
 func initializeWindowAfterCreation(w *glfw.Window) {
 	// TODO: Register NSWindowWillEnterFullScreenNotification and so on.
 	// Enable resizing temporary before making the window fullscreen.
-	C.initializeWindow(C.uintptr_t(w.GetCocoaWindow()))
+	var nswindow = objc.ID(w.GetCocoaWindow())
+	var class_EbitengineWindowDelegate = objc.GetClass("EbitengineWindowDelegate")
+	var sel_delegate = objc.RegisterName("delegate")
+	var delegate = objc.ID(class_EbitengineWindowDelegate).Send(objc.RegisterName("alloc")).Send(objc.RegisterName("initWithOrigDelegate:"), nswindow.Send(sel_delegate))
```
Comment (hajimehoshi) [inline internal/ui/ui_glfw_darwin.go:343]: Let's use `:=` rather than `var`.

File: internal/ui/ui_glfw_darwin.go [hunk 2]
```diff
@@ -345,27 +270,76 @@ func (u *userInterfaceImpl) isNativeFullscreenAvailable() bool {
...
 }
 
+var sel_collectionBehavior = objc.RegisterName("collectionBehavior")
+var sel_setCollectionBehavior = objc.RegisterName("setCollectionBehavior:")
+
 func (u *userInterfaceImpl) setNativeFullscreen(fullscreen bool) {
 	// Toggling fullscreen might ignore events like keyUp. Ensure that events are fired.
 	glfw.WaitEventsTimeout(0.1)
-	C.setNativeFullscreen(C.uintptr_t(u.window.GetCocoaWindow()), C.bool(fullscreen))
+	var window = cocoa.NSWindow{ID: objc.ID(u.window.GetCocoaWindow())}
```
Comment (hajimehoshi) [inline internal/ui/ui_glfw_darwin.go:279]: Let's use `:=` rather than `var`.

File: internal/ui/ui_glfw_darwin.go [hunk 3]
```diff
@@ -345,27 +270,76 @@ func (u *userInterfaceImpl) isNativeFullscreenAvailable() bool {
...
+	if window.StyleMask()&cocoa.NSWindowStyleMaskFullScreen != 0 == fullscreen {
+		return
+	}
+	// Even though EbitengineWindowDelegate is used, this hack is still required.
+	// toggleFullscreen doesn't work when the window is not resizable.
+	var origFullScreen = window.Send(sel_collectionBehavior)&cocoa.NSWindowCollectionBehaviorFullScreenPrimary != 0
```
Comment (hajimehoshi) [inline internal/ui/ui_glfw_darwin.go:285]: Let's use `:=` rather than `var`.

File: internal/ui/ui_glfw_darwin.go [hunk 4]
```diff
@@ -332,11 +239,29 @@ func (u *userInterfaceImpl) nativeWindow() uintptr {
...
 func (u *userInterfaceImpl) isNativeFullscreen() bool {
-	return bool(C.isNativeFullscreen(C.uintptr_t(u.window.GetCocoaWindow())))
+	return cocoa.NSWindow{ID: objc.ID(u.window.GetCocoaWindow())}.StyleMask()&cocoa.NSWindowStyleMaskFullScreen != 0
 }
 
 func (u *userInterfaceImpl) setNativeCursor(shape CursorShape) {
-	C.setNativeCursor(C.int(shape))
+	class_NSCursor := objc.GetClass("NSCursor")
+	NSCursor := objc.ID(class_NSCursor).Send(objc.RegisterName("class"))
+	var sel_performSelector = objc.RegisterName("performSelector:")
```
Comment (hajimehoshi) [inline internal/ui/ui_glfw_darwin.go:248]: Let's use `:=` rather than `var`.

File: internal/cocoa/api_cocoa_darwin.go [hunk 5]
```diff
@@ -38,23 +43,165 @@ var (
...
+	X, Y float64
+}
+
+type CGRect struct {
+	Origin CGPoint
+	Size   CGSize
+}
+
+type NSUInteger = uint
+type NSInteger = int
```
Comment (hajimehoshi) [inline internal/cocoa/api_cocoa_darwin.go:93]: I slightly prefer making them different types

```go
type NSUInteger uint
type NSInteger int
```

What do you think?

File: internal/cocoa/api_cocoa_darwin.go [hunk 6]
```diff
@@ -38,23 +43,165 @@ var (
...
+	sel_screen                             = objc.RegisterName("screen")
+	sel_isVisible                          = objc.RegisterName("isVisible")
+	sel_deviceDescription                  = objc.RegisterName("deviceDescription")
+	sel_objectForKey                       = objc.RegisterName("objectForKey:")
+	sel_unsignedIntValue                   = objc.RegisterName("unsignedIntValue")
 )
 
-type CGFloat float64
+const NSWindowCollectionBehaviorFullScreenPrimary = 1 << 7
+const (
```
Comment (hajimehoshi) [inline internal/cocoa/api_cocoa_darwin.go:72]: Add a space above `const (`

File: internal/cocoa/api_cocoa_darwin.go (hunk 5 above)
Comment (TotallyGamerJet) [reply internal/cocoa/api_cocoa_darwin.go:93]: It was frustrating to forget the cast so doing a type alias was easier. Also, [Apple](https://developer.apple.com/documentation/objectivec/nsuinteger) describes it as a type alias.

Comment (hajimehoshi) [reply internal/cocoa/api_cocoa_darwin.go:93]: I understood.

**Comment by hajimehoshi** [comment]
Do you have a plan how to fix the usage of `go-gl/glfw`?

I'll take a look at this PR later tomorrow. Thank you for your hard work!

**Comment by TotallyGamerJet** [comment]
Yep! I actually started a proof of concept using glfwwin as a basis. It's gonna take quite a while tho since there is a lot of code to port.

File: internal/cocoa/api_cocoa_darwin.go [hunk 7]
```diff
@@ -136,3 +287,39 @@ func (s NSString) String() string {
...
+
+func (screen NSScreen) DeviceDescription() NSDictionary {
+	return NSDictionary{screen.Send(sel_deviceDescription)}
+}
+
+type NSDictionary struct {
+	objc.ID
+}
+
+func (dict NSDictionary) ObjectForKey(object objc.ID) objc.ID {
```
Comment (hajimehoshi) [inline internal/cocoa/api_cocoa_darwin.go:315]: `d`

File: internal/cocoa/api_cocoa_darwin.go [hunk 8]
```diff
@@ -38,23 +43,165 @@ var (
...
+	inv.GetReturnValue(unsafe.Pointer(&color))
+	return color
+}
+
+type NSOperatingSystemVersion struct {
+	Major, Minor, Patch NSInteger
+}
+
+type NSProcessInfo struct {
+	objc.ID
```
Comment (hajimehoshi) [inline internal/cocoa/api_cocoa_darwin.go:125]: Do we have a plan to add other members to `NS` classes? If not, would it make sence to define the type like this?

```go
type NSProcessInfo objc.ID
```

File: internal/ui/ui_glfw_darwin.go [hunk 9]
```diff
@@ -17,224 +17,108 @@
...
+	}), "v@:@")
+	class_EbitengineWindowDelegate.AddMethod(objc.RegisterName("windowWillExitFullScreen:"), objc.NewIMP(func(self objc.ID, _cmd objc.SEL, notification objc.ID) {
+		pushResizableState(self, cocoa.NSNotification{ID: notification}.Object())
+	}), "v@:@")
+	class_EbitengineWindowDelegate.AddMethod(objc.RegisterName("windowDidExitFullScreen:"), objc.NewIMP(func(self objc.ID, _cmd objc.SEL, notification objc.ID) {
+		popResizableState(self, cocoa.NSNotification{ID: notification}.Object())
+		// Do not call setFrame here (#2295). setFrame here causes unexpected results.
+	}), "v@:@")
+	class_EbitengineWindowDelegate.Register()
+
```
Comment (hajimehoshi) [inline internal/ui/ui_glfw_darwin.go:119]: Remove this empty line

File: internal/ui/ui_glfw_darwin.go [hunk 10]
```diff
@@ -317,10 +213,20 @@ func initialMonitorByOS() (*glfw.Monitor, error) {
...
+	pool := cocoa.NSAutoreleasePool_new()
+	screen := cocoa.NSScreen_mainScreen()
+	if window.ID != 0 && window.IsVisibile() {
+		// When the window is visible, the window is already initialized.
+		// [NSScreen mainScreen] sometimes tells a lie when the window is put across monitors (#703).
+		screen = window.Screen()
+	}
+	screenDictionary := screen.DeviceDescription()
+	screenID := cocoa.NSNumber{ID: screenDictionary.ObjectForKey(cocoa.NSString_alloc().InitWithUTF8String("NSScreenNumber").ID)}
+	aID := uintptr(screenID.UnsignedIntValue()) //CGDirectDisplayID
```
Comment (hajimehoshi) [inline internal/ui/ui_glfw_darwin.go:226]: Add a space after `//`

File: internal/ui/ui_glfw_darwin.go [hunk 11]
```diff
@@ -345,27 +269,75 @@ func (u *userInterfaceImpl) isNativeFullscreenAvailable() bool {
...
 
 func (u *userInterfaceImpl) setWindowResizingModeForOS(mode WindowResizingMode) {
 	allowFullscreen := mode == WindowResizingModeOnlyFullscreenEnabled ||
 		mode == WindowResizingModeEnabled
-	C.setAllowFullscreen(C.uintptr_t(u.window.GetCocoaWindow()), C.bool(allowFullscreen))
+	collectionBehavior := int(objc.ID(u.window.GetCocoaWindow()).Send(sel_collectionBehavior))
+	if allowFullscreen {
+		collectionBehavior |= cocoa.NSWindowCollectionBehaviorFullScreenPrimary
+	} else {
+		collectionBehavior &= ^cocoa.NSWindowCollectionBehaviorFullScreenPrimary
```
Comment (hajimehoshi) [inline internal/ui/ui_glfw_darwin.go:331]: `&^=`?

File: internal/ui/ui_glfw_darwin.go [hunk 12]
```diff
@@ -17,224 +17,108 @@
...
+		}
+	}
+	popResizableState := func(self, w objc.ID) {
+		window := cocoa.NSWindow{ID: w}
+		if !getResizable(self) {
+			window.SetStyleMask(window.StyleMask() & ^uint(cocoa.NSWindowStyleMaskResizable))
+		}
+		setResizable(self, false)
+	}
+	class_EbitengineWindowDelegate.AddMethod(objc.RegisterName("initWithOrigDelegate:"), objc.NewIMP(func(self objc.ID, _cmd objc.SEL, origDelegate objc.ID) objc.ID {
```
Comment (hajimehoshi) [inline internal/ui/ui_glfw_darwin.go:70]: Do we need `_` for `cmd`?

Comment (TotallyGamerJet) [reply internal/ui/ui_glfw_darwin.go:70]: Well it's convention but I can remove it

Comment (hajimehoshi) [reply internal/ui/ui_glfw_darwin.go:70]: Let's use `cmd` then. (I'll update the gamepad side)

File: internal/cocoa/api_cocoa_darwin.go (hunk 8 above)
Comment (TotallyGamerJet) [reply internal/cocoa/api_cocoa_darwin.go:125]: If it's embedded then it's possible to call Send without a cast

Comment (hajimehoshi) [reply internal/cocoa/api_cocoa_darwin.go:125]: Ah that makes sense.

**Comment by hajimehoshi** [comment]
By the way, could you answer this question?: https://github.com/hajimehoshi/ebiten/pull/2243#pullrequestreview-1109566958

**Review by hajimehoshi — APPROVED** [review]
LGTM % comments

Thanks!

**Pull request merged**